### ISSUE TITLE
[Discussion] build a linear algebra to cuda compiler

### ISSUE THREAD

**Issue by Jobhdez** [anarchy-ai/LLM-VM#292]
Hello,

The idea of this issue is to initiate a discussion about the a linear algebra to cuda compiler. It would standard linear algebra to sparse linear algebra to cuda.

Maybe @mmirman can start by talking about the purpose for this project. He mentioned on discord in response to a question I had.

thanks

**Comment by mmirman** [comment]
Nah, we need fewer imports and this is an opportunity for us to control our optimizations

**Comment by Jobhdez** [comment]
@mmirman  what use cases do you have in mind? and what are the requirements?

**Comment by mmirman** [comment]
1. not a programming language.  A compiler/interpreter/embedded interpreter for the languages that are currently in use.
2. makes llms faster.

**Comment by mmirman** [comment]
Start with cuda/something, then expand.

**Comment by mmirman** [comment]
If you'd like to scope this, I'm happy to comment and provide bounties, but at the moment scoping this is out of the scope of what I'm working on.

**Comment by Jobhdez** [comment]
Well, its not like I am expert compiler writer but if I were to break this problem into smaller tasks I would make the following tickets/issues:

-  ticket 1: make a parser for a subset of python including the calls to our language models
- ticket 2: convert the dense linear algebra ast into a sparse matrix ast
- ticket 3: convert the sparse matrix ast into the cude ast 
- ticket 4: convert the cuda ast into cuda calls

We can worry about optimizations later.

@mmirman Is this good enough scope? And sorry for the late reply. Let me know please. Thanks.

**Comment by Jobhdez** [comment]
@mmirman I find this problem fascinating and although I am not an expert I am willing to dedicate time into this. Please let me know if you think my scope is good. Im planning to schedule a thread on discord to talk about this. thanks
